lib/datadog/tracing/contrib/roda/instrumentation.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require_relative '../../metadata/ext'
+require_relative 'ext'
+require_relative '../analytics'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module Roda
+        # Instrumentation for Roda
+        module Instrumentation
+          def _roda_handle_main_route
+            instrument(Ext::SPAN_REQUEST) { super }
+          end
+
+          def call
+            instrument(Ext::SPAN_REQUEST) { super }
+          end
+
+          private
+
+          def instrument(span_name, &block)
+            set_distributed_tracing_context!(request.env)
+
+            Tracing.trace(span_name) do |span|
+              begin
+                request_method = request.request_method.to_s.upcase
+
+                span.service = configuration[:service_name]
+                span.span_type = Tracing::Metadata::Ext::HTTP::TYPE_INBOUND
+
+                # Using the http method as a resource, since the URL/path can trigger
+                # a possibly infinite number of resources.
+                span.resource = request_method
+
+                span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_URL, request.path)
+                span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_METHOD, request_method)
+
+                # Add analytics tag to the span
+                if Contrib::Analytics.enabled?(configuration[:analytics_enabled])
+                  Contrib::Analytics.set_sample_rate(span, configuration[:analytics_sample_rate])
+                end
+
+                # Measure service stats
+                Contrib::Analytics.set_measured(span)
+              ensure
+                begin
+                  response = yield
+                rescue
+                  # The status code is unknown to Roda and decided by the upstream web runner.
+                  # In this case, spans default to status code 500 rather than a blank status code.
+                  default_error_status = '500'
+                  span.resource = "#{request_method} #{default_error_status}"
```

Comment:
Possibly strange edge case here where `request.request_method.to_s.upcase` raised an error, hit the `ensure` block, request raised an error, caught by this `rescue` block, then `request_method` will likely be undefined, possibly raising another error.

Reply:
I'm not sure how this error can be thrown because all routes should be defined with a type from the start. Do you have a suggestion for handling this scenario?

Reply:
I paired with Wan and it's not possible for `request.request_method.to_s.upcase` to fail, as it comes straight from Rack and it's part of the Rack contract. This is the same behaviour of our other Rack-ralated integrations (Sinatra, Rails, Hanami, Grace)